src/core/libmaven/mzMassSlicer.cpp

```diff
@@ -237,27 +257,346 @@ void MassSlices::algorithmC(float ppm, float minIntensity, float rtWindow) {
+        if (mavenParameters->stop) {
+            stopSlicing();
+            break;
+        }
+
+        auto firstSlice = *first;
+        if (mzUtils::almostEqual(firstSlice->ionCount, -1.0f))
+            continue;
+
+        for (auto second = next(first); second != end(slices); ++second) {
+            auto secondSlice = *second;
+
+            // stop iterating if the rest of the slices are too far
+            if (firstSlice->mzmax < secondSlice->mzmin)
+                break;
+
+            if (mzUtils::almostEqual(secondSlice->ionCount, -1.0f))
+                continue;
+
+            // check if center of one of the slices lies in the other
+            if ((firstSlice->mz > secondSlice->mzmin
+                 && firstSlice->mz < secondSlice->mzmax
+                 && firstSlice->rt > secondSlice->rtmin
+                 && firstSlice->rt < secondSlice->rtmax)
+                ||
+                (secondSlice->mz > firstSlice->mzmin
+                 && secondSlice->mz < firstSlice->mzmax
+                 && secondSlice->rt > firstSlice->rtmin
+                 && secondSlice->rt < firstSlice->rtmax)) {
+                firstSlice->ionCount = std::max(firstSlice->ionCount,
+                                                secondSlice->ionCount);
+                firstSlice->rtmax = std::max(firstSlice->rtmax,
+                                             secondSlice->rtmax);
+                firstSlice->rtmin = std::min(firstSlice->rtmin,
+                                             secondSlice->rtmin);
+                firstSlice->mzmax = std::max(firstSlice->mzmax,
+                                             secondSlice->mzmax);
+                firstSlice->mzmin = std::min(firstSlice->mzmin,
+                                             secondSlice->mzmin);
+
+                firstSlice->mz = (firstSlice->mzmin + firstSlice->mzmax) / 2.0f;
+                firstSlice->rt = (firstSlice->rtmin + firstSlice->rtmax) / 2.0f;
+                float cutoff = massCutoff->massCutoffValue(firstSlice->mz);
+
+                // make sure that mz window does not get out of control
+                if (firstSlice->mzmin < firstSlice->mz - cutoff)
+                    firstSlice->mzmin =  firstSlice->mz - cutoff;
+                if (firstSlice->mzmax > firstSlice->mz + cutoff)
+                    firstSlice->mzmax =  firstSlice->mz + cutoff;
+
+                // recalculate center mz in case bounds changed
+                firstSlice->mz = (firstSlice->mzmin + firstSlice->mzmax) / 2.0f;
+
+                // flag this slice as already merged, and ignore henceforth
+                secondSlice->ionCount = -1.0f;
+            }
+        }
+        sendSignal("Reducing slices…", first - begin(slices), slices.size());
+    }
+}
+
+void MassSlices::mergeNeighbouringSlices(MassCutoff* massCutoff,
+                                         float rtTolerance)
+{
+    auto slicesInProximity = [&](vector<mzSample*>& samples,
```

Comment:
The whole method is in a lambda function. From what I understand, it's supposed to be used for small blocks of code. It's hurting readability here.

Reply:
I was not aiming for prettiness. While prototyping, we had multiple merge steps, and all of them had a common pattern of iteration and conditionally merging. The iteration part was common and therefore I wrote `_mergeSlices` method for that part. The merging logic was different for each step and that is why the different lambdas were sent to `_mergeSlices` to achieve different kinds of merges. But you are right. Now that we have only one merge step, I should get rid of all this quirky code.